File: ai-solutions/05/composer-1/python/main.py
import os

def solve():
    # Read input file
    input_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '..', '..', 'inputs', '05.txt')
    input_path = os.path.normpath(input_path)
    with open(input_path, 'r') as f:
        input_text = f.read().strip()
    
    # Split input into ranges and available IDs
    parts = input_text.split('\n\n')
    ranges_text = parts[0]
    available_ids_text = parts[1]
    
    # Parse ranges
    ranges = []
    for line in ranges_text.strip().split('\n'):
        start, end = map(int, line.split('-'))
        ranges.append((start, end))
    
    # Parse available IDs
    available_ids = [int(line.strip()) for line in available_ids_text.strip().split('\n')]
    
    # Part 1: Count how many available IDs are fresh
    def part1(ranges, available_ids):
        fresh_count = 0
        for ingredient_id in available_ids:
            is_fresh = False
            for start, end in ranges:
                if start <= ingredient_id <= end:
                    is_fresh = True
                    break
            if is_fresh:
                fresh_count += 1
        return fresh_count
    
    # Part 2: Count total unique IDs covered by ranges
    def part2(ranges):
        # Merge overlapping ranges
        if not ranges:
            return 0
        
        # Sort ranges by start
        sorted_ranges = sorted(ranges, key=lambda x: x[0])
        
        # Merge overlapping ranges
        merged = [sorted_ranges[0]]
        for start, end in sorted_ranges[1:]:
            last_start, last_end = merged[-1]
            if start <= last_end + 1:  # Overlapping or adjacent
                merged[-1] = (last_start, max(last_end, end))
            else:
                merged.append((start, end))
        
        # Count total IDs in merged ranges
        total = 0
        for start, end in merged:
            total += end - start + 1
        
        return total
    
    # Solve both parts
    answer1 = part1(ranges, available_ids)
    answer2 = part2(ranges)
    
    print(answer1)
    print(answer2)

if __name__ == '__main__':
    solve()

File: ai-solutions/05/composer-1/rust/src/main.rs
use std::fs;
use std::path::PathBuf;

fn part1(ranges: &Vec<(i64, i64)>, available_ids: &Vec<i64>) -> usize {
    let mut fresh_count = 0;
    
    for &ingredient_id in available_ids {
        let mut is_fresh = false;
        for &(start, end) in ranges {
            if ingredient_id >= start && ingredient_id <= end {
                is_fresh = true;
                break;
            }
        }
        if is_fresh {
            fresh_count += 1;
        }
    }
    
    fresh_count
}

fn part2(ranges: &Vec<(i64, i64)>) -> i64 {
    if ranges.is_empty() {
        return 0;
    }
    
    // Sort ranges by start
    let mut sorted_ranges = ranges.clone();
    sorted_ranges.sort_by_key(|r| r.0);
    
    // Merge overlapping ranges
    let mut merged = vec![sorted_ranges[0]];
    for &(start, end) in sorted_ranges.iter().skip(1) {
        let last_idx = merged.len() - 1;
        let (last_start, last_end) = merged[last_idx];
        
        if start <= last_end + 1 {  // Overlapping or adjacent
            merged[last_idx] = (last_start, last_end.max(end));
        } else {
            merged.push((start, end));
        }
    }
    
    // Count total IDs in merged ranges
    let mut total = 0;
    for (start, end) in merged {
        total += end - start + 1;
    }
    
    total
}

fn main() {
    // Read input file
    let mut input_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    input_path.pop(); // rust
    input_path.pop(); // composer-1
    input_path.pop(); // 05
    input_path.pop(); // ai-solutions
    input_path.push("inputs");
    input_path.push("05.txt");
    
    let input = fs::read_to_string(input_path)
        .expect("Failed to read input file");
    
    // Split input into ranges and available IDs
    let parts: Vec<&str> = input.trim().split("\n\n").collect();
    let ranges_text = parts[0];
    let available_ids_text = parts[1];
    
    // Parse ranges
    let mut ranges = Vec::new();
    for line in ranges_text.trim().lines() {
        let parts: Vec<&str> = line.split('-').collect();
        let start: i64 = parts[0].parse().expect("Invalid start");
        let end: i64 = parts[1].parse().expect("Invalid end");
        ranges.push((start, end));
    }
    
    // Parse available IDs
    let available_ids: Vec<i64> = available_ids_text
        .trim()
        .lines()
        .map(|line| line.trim().parse().expect("Invalid ID"))
        .collect();
    
    // Solve both parts
    let answer1 = part1(&ranges, &available_ids);
    let answer2 = part2(&ranges);
    
    println!("{}", answer1);
    println!("{}", answer2);
}

